x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/multivalue/MvSort.java

```diff
@@ -0,0 +1,247 @@
+import org.elasticsearch.xpack.esql.expression.function.Param;
+import org.elasticsearch.xpack.esql.planner.PlannerUtils;
+import org.elasticsearch.xpack.esql.type.EsqlDataTypes;
+import org.elasticsearch.xpack.ql.expression.Expression;
+import org.elasticsearch.xpack.ql.expression.Literal;
+import org.elasticsearch.xpack.ql.expression.function.OptionalArgument;
+import org.elasticsearch.xpack.ql.expression.function.scalar.ScalarFunction;
+import org.elasticsearch.xpack.ql.expression.gen.script.ScriptTemplate;
+import org.elasticsearch.xpack.ql.tree.NodeInfo;
+import org.elasticsearch.xpack.ql.tree.Source;
+import org.elasticsearch.xpack.ql.type.DataType;
+import org.elasticsearch.xpack.ql.type.DataTypes;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+import java.util.function.Function;
+
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.FIRST;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.ParamOrdinal.SECOND;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isString;
+import static org.elasticsearch.xpack.ql.expression.TypeResolutions.isType;
+
+/**
+ * Sorts a multivalued field in lexicographical order.
+ */
+public class MvSort extends ScalarFunction implements OptionalArgument, EvaluatorMapper {
+    private final Expression field, order;
+
+    private static final Literal ASC = new Literal(Source.EMPTY, "ASC", DataTypes.TEXT);
+
+    private static final Literal DESC = new Literal(Source.EMPTY, "DESC", DataTypes.TEXT);
+
+    @FunctionInfo(
+        returnType = { "boolean", "date", "double", "integer", "ip", "keyword", "long", "text", "version" },
+        description = "Sorts a multivalued field in lexicographical order."
+    )
+    public MvSort(
+        Source source,
+        @Param(
+            name = "field",
+            type = { "boolean", "date", "double", "integer", "ip", "keyword", "long", "text", "version" },
+            description = "A multivalued field"
+        ) Expression field,
+        @Param(name = "order", type = { "keyword" }, description = "sort order", optional = true) Expression order
+    ) {
+        super(source, order == null ? Arrays.asList(field, ASC) : Arrays.asList(field, order));
+        this.field = field;
+        this.order = order == null ? ASC : order;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        TypeResolution resolution = isType(field, EsqlDataTypes::isRepresentable, sourceText(), FIRST, "representable");
+
+        if (resolution.unresolved()) {
+            return resolution;
+        }
+
+        if (order != null) {
+            resolution = isString(order, sourceText(), SECOND);
+            if (resolution.unresolved()) {
+                return resolution;
+            }
+        }
```

Comment:
Should we force this `order` parameter to be a literal? It feels odd to chose the order based on another field.

Reply:
I'll defer the foldable check at the end of logical planner, similar to auto_bucket, it feels odd to choose the order from another field, but it might from a eval reference?